File: macos/Sources/GlyphstrokeMac/GestureEngine.swift
import Foundation
import AppKit
import GlyphstrokeCore

/// Связка всего: перехват мыши, след, распознавание, выполнение.
///
/// Здесь же принимается главное решение каждого росчерка — забрать щелчок себе
/// или отдать приложению. Правило то же, что в версии для Linux: щелчок
/// забирается, только если жест действительно выполнен. Коротким движением
/// человек хотел открыть контекстное меню, а неопознанный росчерк — это
/// промах, и молча съедать его нельзя: пользователь решит, что мышь сломалась.
public final class GestureEngine: MouseTapDelegate {
    private let store: Store
    private let tap = MouseTap()
    private let trail: TrailWindow
    private let actions = ActionRunner()

    private var settings: Settings
    private var gestures: [Gesture] = []
    private var recognizer = Recognizer(gestures: [])

    private var stroke: [Point] = []
    /// Приложение, которое было впереди в момент нажатия. Спрашиваем один раз:
    /// пока рисуется росчерк, оно поменяться не может, а опрос не бесплатный.
    private var strokeApp: String?

    /// Куда писать происходящее — в журнал приложения и в окно проверки.
    public var onLog: ((String) -> Void)?
    /// Что распознали: имя жеста (или `nil`) и код направлений.
    public var onRecognized: ((Match) -> Void)?

    public private(set) var paused = false {
        didSet {
            tap.paused = paused
            if paused { trail.cancel() }
        }
    }

    public init(store: Store = Store()) {
        self.store = store
        self.settings = store.loadSettings()
        self.trail = TrailWindow(settings: settings.overlay)
        tap.delegate = self
        actions.onLog = { [weak self] message in self?.log(message) }
        reload()
    }

    /// Перечитать настройки и жесты с диска.
    public func reload() {
        settings = store.loadSettings()
        gestures = store.loadGestures()
        trail.settings = settings.overlay
        tap.triggerButton = settings.triggerButton
        rebuildRecognizer(for: nil)
        log("жестов загружено: \(gestures.filter(\.enabled).count)")
    }

    public func start() throws {
        try tap.start()
        log("перехват включён, кнопка-модификатор: \(settings.triggerButton)")
    }

    public func stop() {
        tap.stop()
        trail.cancel()
    }

    public func setPaused(_ value: Bool) {
        paused = value
        WindowContext.forget()
        log(value ? "перехват на паузе" : "перехват продолжен")
    }

    public var isRunning: Bool { tap.isRunning }
    public var gestureCount: Int { gestures.filter(\.enabled).count }

    // MARK: - Перехватчик

    public func tapDidBeginStroke(at point: Point) {
        strokeApp = WindowContext.current()

        // Исключённые программы: в них мышь не трогаем вовсе. Проверка именно
        // здесь, а не при выполнении, — иначе в чужом окне пропадал бы щелчок.
        if isExcluded(app: strokeApp) {
            stroke = []
            return
        }
        if settings.pauseInFullscreen && WindowContext.isFrontmostFullscreen() {
            stroke = []
            return
        }

        stroke = [point]
        rebuildRecognizer(for: strokeApp)
        trail.begin(at: point)
    }

    public func tapDidExtendStroke(to point: Point) {
        guard !stroke.isEmpty else { return }
        stroke.append(point)
        trail.extend(to: point)
    }

    public func tapDidEndStroke(at point: Point) -> Bool {
        defer {
            trail.end()
            stroke = []
        }
        guard !stroke.isEmpty else { return false }
        stroke.append(point)

        let length = Geometry.pathLength(stroke)
        if length < settings.minStrokePx {
            // Это не росчерк, а обычный щелчок: пусть уходит приложению.
            return false
        }

        let match = recognizer.recognize(stroke)
        onRecognized?(match)

        guard let name = match.name,
              let gesture = gestures.first(where: { $0.enabled && $0.name == name }) else {
            log("не опознано: \(match.code)"
                + (match.runnerUp.map { ", ближе всех «\($0)» (\(percent(match.runnerUpScore)))" } ?? ""))
            // Неопознанный росчерк по умолчанию отдаём приложению целиком.
            return settings.unrecognized == "swallow"
        }

        log("жест «\(gesture.name)» (\(match.code), \(percent(match.score)))")
        actions.run(gesture.actions)
        return true
    }

    // MARK: - Внутреннее

    /// Собрать распознаватель из жестов, подходящих текущей программе.
    ///
    /// Отбор до распознавания, а не после, — не ради скорости: жест «вниз» в
    /// браузере и жест «вниз» в терминале могут быть разными, и лишний
    /// кандидат портил бы отрыв от второго места, из-за чего не сработал бы ни
    /// один из двух.
    private func rebuildRecognizer(for app: String?) {
        let usable = gestures.filter { $0.enabled && $0.event.isEmpty && $0.matches(app: app) }
        recognizer = Recognizer(gestures: usable.map(\.definition),
                                minScore: settings.minScore,
                                minMargin: settings.minMargin)
    }

    private func isExcluded(app: String?) -> Bool {
        guard let app, !settings.excludedApps.isEmpty else { return false }
        return settings.excludedApps.contains { Gesture.appMatches(pattern: $0, app: app) }
    }

    private func percent(_ score: Double) -> String {
        String(format: "%.0f%%", score * 100)
    }

    private func log(_ message: String) {
        onLog?(message)
    }
}

File: windows/src/Glyphstroke.App/GestureEngine.cs
using Glyphstroke.Core;

namespace Glyphstroke.App;

/// <summary>Связка всего: перехват мыши, след, распознавание, выполнение.</summary>
/// <remarks>
/// Здесь принимается главное решение каждого росчерка — забрать щелчок себе или
/// отдать программе. Правило то же, что в версиях для Linux и macOS: щелчок
/// забирается, только если жест действительно выполнен. Коротким движением
/// человек хотел открыть контекстное меню, а неопознанный росчерк — это промах,
/// и молча съедать его нельзя: пользователь решит, что мышь сломалась.
/// </remarks>
public sealed class GestureEngine : IMouseHookListener, IDisposable
{
    private readonly Store _store;
    private readonly MouseHook _hook;
    private readonly TrailWindow _trail;
    private readonly ActionRunner _actions = new();

    private Settings _settings;
    private List<Gesture> _gestures = new();
    private Recognizer _recognizer = new(Array.Empty<GestureDefinition>());

    private readonly List<Point> _stroke = new();
    private string? _strokeApp;
    private IntPtr _targetWindow;   // окно под началом росчерка — цель действий

    public Action<string>? OnLog { get; set; }
    public Action<Match>? OnRecognized { get; set; }

    public GestureEngine(Store store, TrailWindow trail)
    {
        _store = store;
        _trail = trail;
        _settings = store.LoadSettings();
        _hook = new MouseHook(this);
        _actions.OnLog = message => OnLog?.Invoke(message);
        Reload();
    }

    public bool Paused
    {
        get => _hook.Paused;
        set
        {
            _hook.Paused = value;
            if (value)
            {
                _trail.Cancel();
            }
            WindowContext.Forget();
            Log(value ? "перехват на паузе" : "перехват продолжен");
        }
    }

    public bool IsRunning => _hook.IsRunning;

    public int GestureCount => _gestures.Count(gesture => gesture.Enabled);

    public Settings Settings => _settings;

    /// <summary>Перечитать настройки и жесты с диска.</summary>
    public void Reload()
    {
        _settings = _store.LoadSettings();
        _gestures = _store.LoadGestures();
        _trail.Settings = _settings.Overlay;
        _hook.TriggerButton = _settings.TriggerButton;
        Rebuild(null);
        Log($"жестов загружено: {GestureCount}");
    }

    public void Start()
    {
        _hook.Start();
        Log($"перехват включён, кнопка-модификатор: {_settings.TriggerButton}");
    }

    public void Stop()
    {
        _hook.Stop();
        _trail.Cancel();
    }

    public void Dispose() => Stop();

    // --- перехватчик ---

    public bool StrokeBegan(Point point)
    {
        _stroke.Clear();
        _strokeApp = WindowContext.Current();
        // Цель действий — окно под точкой начала росчерка, а не активное окно.
        _targetWindow = WindowContext.WindowUnder((int)point.X, (int)point.Y);

        // Исключённые и полноэкранные — не наш случай: возвращаем false, и
        // перехватчик отдаёт нажатие программе как есть (без глотания и подмены).
        if (IsExcluded(_strokeApp))
        {
            return false;
        }
        if (_settings.PauseInFullscreen && WindowContext.IsFullscreen())
        {
            return false;
        }

        _stroke.Add(point);
        Rebuild(_strokeApp);
        // НЕ блокируем колбэк хука: рисуем след асинхронно.
        _trail.Dispatcher.InvokeAsync(() => _trail.Begin(point));
        return true;
    }

    public void StrokeExtended(Point point)
    {
        if (_stroke.Count == 0)
        {
            return;
        }
        _stroke.Add(point);
        _trail.Dispatcher.InvokeAsync(() => _trail.Extend(point));
    }

    public bool StrokeEnded(Point point)
    {
        if (_stroke.Count == 0)
        {
            return false;
        }
        _stroke.Add(point);
        _trail.Dispatcher.InvokeAsync(() => _trail.End());

        double length = Geometry.PathLength(_stroke);
        var stroke = _stroke.ToList();
        _stroke.Clear();

        if (length < _settings.MinStrokePx)
        {
            // Это не росчерк, а обычный щелчок: пусть уходит программе.
            return false;
        }

        var match = _recognizer.Recognize(stroke);
        OnRecognized?.Invoke(match);

        var gesture = match.Name is null
            ? null
            : _gestures.FirstOrDefault(item => item.Enabled && item.Name == match.Name);

        if (gesture is null)
        {
            string closest = match.RunnerUp is null
                ? string.Empty
                : $", ближе всех «{match.RunnerUp}» ({match.RunnerUpScore * 100:0}%)";
            Log($"не опознано: {match.Code}{closest}");
            // Неопознанный росчерк по умолчанию отдаём программе целиком.
            return string.Equals(_settings.Unrecognized, "swallow", StringComparison.OrdinalIgnoreCase);
        }

        Log($"жест «{gesture.Name}» ({match.Code}, {match.Score * 100:0}%)");
        if (gesture.Menu.Count > 0)
        {
            // Есть пункты — вместо своих действий жест открывает меню у курсора.
            Native.GetCursorPos(out var at);
            var items = gesture.Menu.ToList();
            int timeout = _settings.MenuTimeoutMs;
            var target = _targetWindow;
            _trail.Dispatcher.InvokeAsync(() =>
            {
                var overlay = new MenuOverlay(items, at, timeout, item => _actions.Run(item.Actions, target));
                overlay.Show();
                overlay.Activate();
            });
            return true;
        }
        // Действия — в фоновый поток: они могут запускать программы и слать
        // ввод, и держать на этом колбэк хука нельзя (замёрзнет мышь всей системы).
        var actions = gesture.Actions.ToList();
        var target = _targetWindow;
        Task.Run(() => _actions.Run(actions, target));
        return true;
    }

    // --- внутреннее ---

    /// <summary>Собрать распознаватель из жестов, подходящих текущей программе.</summary>
    /// <remarks>
    /// Отбор до распознавания, а не после, — не ради скорости: жест «вниз» в
    /// браузере и жест «вниз» в терминале могут быть разными, и лишний кандидат
    /// портил бы отрыв от второго места, из-за чего не сработал бы ни один.
    /// </remarks>
    private void Rebuild(string? app)
    {
        var usable = _gestures
            .Where(gesture => gesture.Enabled && gesture.Event.Length == 0 && gesture.Matches(app))
            .Select(gesture => gesture.ToDefinition())
            .ToList();
        _recognizer = new Recognizer(usable, _settings.MinScore, _settings.MinMargin);
    }

    private bool IsExcluded(string? app)
    {
        if (app is null || _settings.ExcludedApps.Count == 0)
        {
            return false;
        }
        return _settings.ExcludedApps.Any(pattern => Gesture.AppMatches(pattern, app));
    }

    private void Log(string message) => OnLog?.Invoke(message);
}
